### ISSUE TITLE
maximal number of DLLs reached

### ISSUE THREAD

**Issue by wenbostar** [rstudio/rstudio#2375]
### System details

    RStudio Edition : Desktop
    RStudio Version : 1.1.423  
    OS Version : Mac OS X 10_13_3
    R Version : 3.4.3 (2017-11-30) 

### Steps to reproduce the problem
> library(APEAK)
Error: package or namespace load failed for ‘APEAK’ in dyn.load(file, DLLpath = DLLpath, ...):
 unable to load shared object '/usr/local/lib/R/3.4/site-library/mclust/libs/mclust.so':
  `maximal number of DLLs reached...

> head(x)
Error in dyn.load(file, DLLpath = DLLpath, ...) : 
  unable to load shared object '/usr/local/lib/R/3.4/site-library/utf8/libs/utf8.so':
  `maximal number of DLLs reached...

I only got the above error when I ran "library(APEAK)" in Rstudio. There is no error when I run in Mac terminal.

```
 $ R

R version 3.4.3 (2017-11-30) -- "Kite-Eating Tree"
Copyright (C) 2017 The R Foundation for Statistical Computing
Platform: x86_64-apple-darwin17.3.0 (64-bit)

R is free software and comes with ABSOLUTELY NO WARRANTY.
You are welcome to redistribute it under certain conditions.
Type 'license()' or 'licence()' for distribution details.

  Natural language support but running in an English locale

R is a collaborative project with many contributors.
Type 'contributors()' for more information and
'citation()' on how to cite R or R packages in publications.

Type 'demo()' for some demos, 'help()' for on-line help, or
'help.start()' for an HTML browser interface to help.
Type 'q()' to quit R.

During startup - Warning messages:
1: Setting LC_COLLATE failed, using "C"
2: Setting LC_TIME failed, using "C"
3: Setting LC_MESSAGES failed, using "C"
4: Setting LC_MONETARY failed, using "C"
> library(APEAK)
Loading required package: VennDiagram
Loading required package: grid
Loading required package: futile.logger
Loading required package: pROC
Type 'citation("pROC")' for a citation.

Attaching package: ‘pROC’

The following objects are masked from ‘package:stats’:

    cov, smooth, var

Loading required package: SSPA
Loading required package: qvalue
Loading required package: lattice
Loading required package: limma
>
```

Does anyone know how to fix this problem?

**Comment by ronblum** [comment]
@wenbostar Do you have this issue with any other packages? APEAK isn't in the CRAN package repository.

**Comment by wenbostar** [comment]
@ronblum , thanks. I don't have this issue with other packages which I have used so far. APEAK is a package I developed and it depends on a lot of packages. In Linux system, when I get the problem "maximal number of DLLs reached", I can fix that with adding "export R_MAX_NUM_DLLS=1000" into the .bashrc file. In mac system, I can also add "export R_MAX_NUM_DLLS=1000" into the .bashrc file and you can see that there is no error when I run in Mac terminal. The error only occurred in RStudio.

**Comment by tungmilan** [comment]
I also had this problem with `leaflet` package if I knitted my .Rmd file too many times. The only solution was to exit RStudio and started again. The problem would return again after knitting several times.

**Comment by ronblum** [comment]
@tungmilan Can you please provide a reproduce example?

**Comment by tungmilan** [comment]
@ronblum : I attached the file. 

[leaflet_max_dll.Rmd.txt](https://github.com/rstudio/rstudio/files/1808450/leaflet_max_dll.Rmd.txt)

**Comment by ronblum** [comment]
@tungmilan After installing `USGS-R/hydroMap` and `ggmap`, plus adding `always_allow_html: yes`, I was able to successfully knit the file. However, after about 30 times, I still wasn't able to reproduce the problem.

So, @wenbostar and @tungmilan, this is tough—we seem to have the same environment and product setup, yet we’re getting different results. I’m going to leave this issue open so we can continue tracking it. If you come up with anything else that could be useful, that would be great. And this way, if any other users experience a similar problem, they might be able to provide additional information that could help us identify the cause and address the issue.

**Comment by tungmilan** [comment]
@kevinushey : you were probably right. My example Rmd is part of a bigger Rmd file in which there are about 40 packages used (not counting dependencies)

**Comment by bio-la** [comment]
hi, I have the same problem and I found it's not really dependent on which package you're trying to load, but on the memory footprint of the R session- and how much memory is occupied in general. I had this error coming up with a list of packages which indeed have a lot of dependencies (like zinbwave, Seurat, scater, scran) the problem is only in Rstudio, while setting the R_MAX_NUM_DLLS seems to have an effect only on R from the terminal.

The error doesn't show up immediately after loading a faulty package - if there's such a thing like a package faulty of having a large footprint really-  but will occur randomly during the session, I assume once the threshold is met, which can be after loading any package or after having loaded a large matrix. 

I realize that the factors impacting on this are a lot but I can't see how to fix this - I don't think that limiting myself to load only a certain amount of packages and data at the time is a solution that I can apply to everything, in particular in cases in which I am actually building something from scratch.

once the error comes up, i can't even get sessionInfo so most of the times I have to quit the session and restart again, only to find the problem will be there, again randomly. for example, i just started a session loading what I would define a normal set of packages for my usual analyses 
(here,dplyr,Matrix,scater,scran,reshape2,fields,ggplot2,SDAtools,wesanderson,ggrepel) and my enviroment only stores a 24 mb list. the error this time appeared when I was trying to load a large seurat object (loaded in a separate session is 3.9 gb). Honestly, I've done worse than that.

again i wish I could do a sessionInfo but it's not working as I said - I get 

sessionInfo()
Error in La_library() : LAPACK routines cannot be loaded
In addition: Warning message:
In La_library() :
  unable to load shared object '/Library/Frameworks/R.framework/Resources/modules//lapack.so':
  `maximal number of DLLs reached... 

happy to give further info on my computer and R version and so on, if anyone can tell me what i'm doing wrong, or what to look for...

**Comment by kfeinauer** [comment]
@bio-la Have you tried exporting the environment variable in a terminal and then launching RStudio that way as well?

For example:

```
export R_MAX_NUM_DLLS=100
/Applications/RStudio.app/Contents/MacOS/RStudio 
```

When I tried this, I noticed my RStudio environment was limited to loading 100 packages. As long as the variable is set before R is started, I think it should work correctly.

**Comment by wenbostar** [comment]
@bio-la , it looks like the problem has been fixed in the latest version of R (version R 3.5.0).

https://cran.r-project.org/doc/manuals/r-release/NEWS.html
> The maximum number of DLLs that can be loaded into R e.g. via dyn.load() has been increased up to 614 when the OS limit on the number of open files allows.

**Comment by bio-la** [comment]
thanks @kfeinauer I did try that. the maximum number I tested was 500 but it seemed to work only in the R session from terminal, i'll try to launch Rstudio from the terminal after exporting the variable- didn't think that would make a difference.

@wenbostar didn't install 3.5 yet, (i'm on 3.4.3/ .4 ) but will definitely give it a go soon. good to know that the problem was somehow addressed in this version. thanks!
